MarkupsToModel/Logic/vtkSlicerMarkupsToModelLogic.cxx

```diff
@@ -538,6 +577,373 @@ void vtkSlicerMarkupsToModelLogic::UpdateOutputHermiteSplineModel(vtkMRMLMarkups
+
+  // dependent values
+  const int numDimensions = 3; // this should never be changed from 3
+  int numDependentValues = numPoints * numDimensions;
+  std::vector<double> dependentValues(numDependentValues); // dependent values
+  dependentValues.assign(numDependentValues, 0.0);
+  for (int p = 0; p < numPoints; p++) // p = point index
+  {
+    double* currentPoint = points->GetPoint(p);
+    for (int d = 0; d < numDimensions; d++) // d = dimension index
+    {
+      double value = currentPoint[d];
+      dependentValues[p * numDimensions + d] = value;
+    }
+  }
+  std::vector<double*> dependentMatrix(numPoints);
+  dependentMatrix.assign(numPoints, NULL);
+  for (int p = 0; p < numPoints; p++)
+  {
+    dependentMatrix[p] = &(dependentValues[p * numDimensions]);
+  }
+  double** dependentMatrixPtr = &(dependentMatrix[0]);
+
+  // solution to least squares
+  int totalNumberCoefficients = numDimensions*numPolynomialCoefficients;
+  std::vector<double> coefficientValues(totalNumberCoefficients);
+  coefficientValues.assign(totalNumberCoefficients, 0.0);
+  std::vector<double*> coefficientMatrix(numPolynomialCoefficients);
+  for (int c = 0; c < numPolynomialCoefficients; c++)
+  {
+    coefficientMatrix[c] = &(coefficientValues[c * numDimensions]);
+  }
+  double** coefficientMatrixPtr = &(coefficientMatrix[0]); // the solution
+
+  // Input the forumulation into SolveLeastSquares
+  vtkMath::SolveLeastSquares(numPoints, independentMatrixPtr, numPolynomialCoefficients, dependentMatrixPtr, numDimensions, coefficientMatrixPtr);
+
+  // Use the values to generate points along the polynomial curve
+  vtkSmartPointer<vtkPoints> smoothedPoints = vtkSmartPointer<vtkPoints>::New(); // points
+  vtkSmartPointer< vtkCellArray > smoothedLines = vtkSmartPointer<  vtkCellArray >::New(); // lines
+  int numControlPointsOnTube = numPoints * markupsToModelModuleNode->GetTubeSamplingFrequency();
+  for (int p = 0; p < numControlPointsOnTube; p++) // p = point index
+  {
+    double pointMm[3];
+    for (int d = 0; d < numDimensions; d++)
+    {
+      pointMm[d] = 0.0;
+      for (int c = 0; c < numPolynomialCoefficients; c++)
+      {
+        double coefficient = coefficientValues[c * numDimensions + d];
+        pointMm[d] += coefficient * std::pow( (double(p)/(numControlPointsOnTube-1)), c );
+      }
+    }
+    smoothedPoints->InsertPoint(p, pointMm);
+    // now connect to the previous point as a line
+    if (p > 0)
+    {
+      smoothedLines->InsertNextCell(2);
+      smoothedLines->InsertCellPoint(p-1);
+      smoothedLines->InsertCellPoint(p);
+    }
+  }
+
+  // Convert the points to a tube model
+  vtkSmartPointer< vtkPolyData >smoothedSegments = vtkSmartPointer< vtkPolyData >::New();
+  smoothedSegments->Initialize();
+  smoothedSegments->SetPoints(smoothedPoints);
+  smoothedSegments->SetLines(smoothedLines);
+  
+  vtkSmartPointer< vtkTubeFilter> tubeFilter = vtkSmartPointer< vtkTubeFilter>::New();
+#if (VTK_MAJOR_VERSION <= 5)
```

Comment:
Slicer is no more backward-compatible with VTK5, so you can remove VTK_MAJOR_VERSION<=5 branches